tools/autograd/utils.py

```diff
@@ -74,6 +74,12 @@ def is_tensor_method(declaration):
 def is_out_variant(decl):
     return decl['name'].endswith('_out')
 
+def op_name_with_overload(decl):
+    name = decl['operator_name_with_overload']
+    if is_out_variant(decl):
+        name = decl['operator_name_with_overload'][:-len('_out')]
```

Comment:
This looks weird - 'operator_name_with_overload' is the main name + overload name string, while is_out_variant() is to check whether the main name ends with '_out' - so looks like it removes the last 4 chars from overload part if main name ends with '_out'...

Here is an example:
name: eye_out
overload_name: m_out
schema_string: aten::eye.m_out(int n, int m, *, Tensor(a!) out) -> Tensor(a!)

operator_name_with_overload would be 'aten::eye.m_out'
and this method would return 'aten::eye.m'

Reply:
Good catch! I'm curious to know what the rationale behind removing the out-variant (stripping out the trailing `_out`) is here?

Reply:
@ljk53 I'm seeing that in Declarations.yaml (by convention?) out variants have an overload name of `.out`. For example:

```
- name: abs_out
  operator_name: abs
  overload_name: out
  use_c10_dispatcher: with_codegenerated_unboxing_wrapper
  manual_kernel_registration: false
  category_override: ''
  matches_jit_signature: true
  schema_string: aten::abs.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
```

and:

```
- name: polar_out
  operator_name: polar
  overload_name: out
  use_c10_dispatcher: with_codegenerated_unboxing_wrapper
  manual_kernel_registration: false
  category_override: ''
  matches_jit_signature: true
  schema_string: aten::polar.out(Tensor abs, Tensor angle, *, Tensor(a!) out) -> Tensor(a!)
```

Q. Was the original intention to keep all the overloads with an `.out` suffix if the base operator is included in the selective build?

i.e. if `abc`, `bcd`, and `xyz` are selected, then also select `abc.out`, `bcd.out`, and `xyz.out`?

Reply:
@ljk53 Okay - I think I can remove this logic altogether and just return the full operator name with overload, and it should be fine. lmk what you think.

Reply:
I think the main operator name always has "_out" suffix if it's an out-variant - but the overload part might not always be ".out", (eye_out is an counterexample, its overload is m_out). I think it's fine to apply [:-len('_out')] on the main name, but applying it on the concatenated name `operator_name_with_overload` seems confusing...

I'm not 100% sure about the original intention of this conversion - maybe if an op is selected, then its base and out-variant can both be called and should include both? We should dig into it later :)

Reply:
> maybe if an op is selected, then its base and out-variant can both be called and should include both? We should dig into it later :)

I also think this is the main reason for doing so. **If** that is the case, then removing the check entirely should be fine because:
1. In case `include_all_overloads` is true, the build will pick up all overloads (even the ones with the `.out` variant)
2. When we switch to tracing for training, we will know all the operator names (with overload), and then we don't need to over-select ops just for training purposes. Till such time that we don't do tracing for training, [1] should cover the case - thoughts?

Reply:
@ljk53 So for `name: eye_out`, there are 2 overloads, one with `out` and one with `m_out`, so previously when removing the `_out` suffix from base operator name, things seem fine, but doing this with the overload is clearly not okay like you mentioned. I believe `include_all_overloads=True` should include both overloads (which is what the earlier code was trying to do?)

Reply:
ok, I took a closer took - seems it's because the Declarations.yaml and native_functions.yaml have slightly different naming convention. In native functions.yaml, the out-place version and base version have the same 'name', e.g.:
```
- func: abs(Tensor self) -> Tensor
- func: abs.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
```

In Declarations.yaml, they have different 'name':
```
- name: abs
  operator_name: abs
  overload_name: ''
- name: abs_out
  operator_name: abs
  overload_name: out
```

The unboxing_wrapper codegen operates on Declarations.yaml, but the allowlist ops are in native functions 'name' format. E.g. root_ops = ['aten::abs.out"], after stripping out the namespace and the overload part, we get 'abs', which still won't match the `decl['name']` - I think that's why the old logic has to remove the '_out' suffix in the 'name' field.

`operator_name_with_overload` seems to be generated from native_functions.yaml schema string, i.e. the 'name' part doesn't have the '_out' suffix, so this step doesn't seem to be necessary.

Reply:
@ljk53 I agree - and am wondering if it would have been simpler for the unboxing/autograd codegen to look at `operator_name` instead of `name` (since `operator_name` is what maps to the name in `native_functions.yaml`, and not `name`, which seems to be something else).